File: 프로그래머스/2/340211. ［PCCP 기출문제］ 3번 ／ 충돌위험 찾기/［PCCP 기출문제］ 3번 ／ 충돌위험 찾기.py
def solution(points, routes):
    answer = 0
    

    field = [[[] for i in range(101)] for j in range(101)]
    robots = []
    
    for route in routes:
        start = route[0]
        dest = route[-1]
        robot = {
            "start" : points[start-1][:], 
            "end" : points[route[1]-1][:], 
            "pos" : points[start-1][:],
            "dest" : points[dest-1][:],
            "visit" : route[2:]
        }
        
        robots.append(robot)


    while len(robots) > 0:

        arrive = []
        collision = []
        for i in range(len(robots)):
            robot = robots[i]
            pos = robot["pos"]
            start = robot["start"]
            end = robot["end"]
            dest = robot["dest"]
            visit = robot["visit"]
            
            if (len(field[pos[0]][pos[1]]) == 1):
                collision.append(pos[:])
            field[pos[0]][pos[1]].append(robot)
            
            if (pos[0] == end[0] and pos[1] == end[1] and len(visit) == 0):
                arrive.append(i)
                           
            else:
                if (pos[0] == end[0] and pos[1] == end[1]): 
                    
                    end[0] = points[visit[0]-1][0]
                    end[1] = points[visit[0]-1][1]
                    del visit[0]
                                   
                dr = 0
                dc = 0
                if (end[0] > pos[0]):
                    dr = 1
                elif (end[0] < pos[0]):
                    dr = -1
                if (end[1] > pos[1]):
                    dc = 1
                elif (end[1] < pos[1]):
                    dc = -1
                if (pos[0] == end[0]):
                    pos[1] += dc
                else:
                    pos[0] += dr
                
                
        answer += len(collision)
        
        for robot in robots:
            field[robot["pos"][0]][robot["pos"][1]] = []
        
        for i in range(len(arrive)-1, -1, -1):
            arr = arrive[i]
            del robots[arr]
        
        
            
    return answer

File: 프로그래머스/2/340211. ［PCCP 기출문제］ 3번 ／ 충돌위험 찾기/［PCCP 기출문제］ 3번 ／ 충돌위험 찾기.js
function solution(points, routes) {
    let answer = 0;
        
    let current = [];
    
    for (let i = 0; i < routes.length; i++) {
        current.push([...points[routes[i][0]-1]]);
    }
    
    let flag = [];
    for (let i = 0; i < routes.length; i++) {
        flag[i] = [];
        for (let j = 0; j < routes[i].length; j++) {
            flag[i][j] = false;
        }
    }
    
    let count = 0;
    
    while (current.length > 0) {
        // 현재 위치에서 충돌이 있는지 체크
        // console.log(current);
        // console.log(flag);
        let check = [];
        count++;
        for (let i = 0; i < current.length-1; i++) {
            if (check[i]) continue;
            if (flag[i][flag[i].length-1]) continue;
            for (let j = i+1; j < current.length; j++) {
                if (flag[j][flag[j].length-1]) continue;
                if (current[i][0] === current[j][0] && current[i][1] === current[j][1]) {
                    check[i] = true;
                    check[j] = true;
                }
            }
            if (check[i]) {
            
                // console.log(i);
                // console.log(check);
                
                answer++
            }
        }
        
        
        // 다 이동했는지 체크
        let finish = true;
        let temp = [];
        let temp2 = [];
        for (let i = 0; i < flag.length; i++) {
            if (!flag[i][flag[i].length-1]) {
                finish = false;

                break;
            }
        }
        // current = temp;
        // flag = temp2;
        if (finish) break;
        
        // 루트를 확인하며 이동
        for (let i = 0; i < flag.length; i++) {
            for (let j = 0; j < flag[i].length; j++) {
                if (!flag[i][j]) {
                    let target = points[routes[i][j]-1];
                    if (current[i][0] < target[0]) current[i][0]++;
                    else if (current[i][0] > target[0]) current[i][0]--;
                    else if (current[i][1] < target[1]) current[i][1]++;
                    else if (current[i][1] > target[1]) current[i][1]--;
                    else {
                        flag[i][j] = true;
                        continue;
                    }
                    break;
                }
            }
        }
    }
    
    return answer;
}